### PR TITLE
feat(daemon): clamp oversized inline media on the prompt path

### PR DESCRIPTION
**Pull request by doudouOUC**
## Summary

- Add `clampInlineMediaPart` utility that replaces inline image/audio/blob payloads exceeding a configurable byte ceiling (`QWEN_CODE_MAX_INLINE_MEDIA_BYTES`, default 10 MB) with a sanitized text placeholder, preventing oversized daemon media from blowing up request size or token budget.
- Wire the clamp into `Session.#resolvePrompt` and advertise `audio: true` in HTTP daemon `promptCapabilities`.
- Fix a type narrowing issue where `fileData.fileUri` became optional after clamping.

## Test plan

- [ ] Unit tests pass for `inlineMediaLimit.ts` (covers base64/buffer thresholds, audio/image/blob, env-var override)
- [ ] Session integration test verifies oversized inline parts are replaced with placeholder text
- [ ] Existing daemon E2E tests still pass

🤖 Generated with [Qwen Code](https://github.com/QwenLM/qwen-code)

### CODE REVIEW

File: packages/core/src/core/inlineMediaLimit.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      start = commaIndex + 1;
+    }
+  }
+  const length = base64.length - start;
+  if (length === 0) {
+    return 0;
+  }
+  // Padding chars are always trailing, so endsWith on the full string is safe.
+  const padding = base64.endsWith('==') ? 2 : base64.endsWith('=') ? 1 : 0;
+  return Math.floor((length * 3) / 4) - padding;
```
[the remark this answers is not in the document]
Comment (doudouOUC) [reply packages/core/src/core/inlineMediaLimit.ts:54]: Thanks — won't take this one. A negative return for degenerate inputs like bare `=` is harmless: the caller (`clampInlineMediaPart`) compares `bytes <= limitBytes`, so a negative value passes the part through unchanged — correct behavior for malformed/empty data. Real inline media from MCP servers and file reads is always valid base64, so this path is unreachable in production.

File: packages/cli/src/acp-integration/session/Session.ts [hunk 2]
```diff
@@ -2956,12 +2957,14 @@ export class Session implements SessionContext {
...
       // Type guard for blob resources
       if ('blob' in contextPart && contextPart.blob) {
-        processedQueryParts.push({
-          inlineData: {
-            mimeType: contextPart.mimeType ?? 'application/octet-stream',
-            data: contextPart.blob,
-          },
-        });
+        processedQueryParts.push(
+          clampInlineMediaPart({
```
Comment (wenshao) [inline packages/cli/src/acp-integration/session/Session.ts:2961]: **[Suggestion]** No integration test covers this blob branch — the second `clampInlineMediaPart` call site. The new test ("degrades an oversized inline image...") exercises only the `image`/`audio` switch arm above. A search of `Session.test.ts` shows zero tests using `embeddedContext` or `blob` resource parts.

This call site has its own distinct behavior: `contextPart.mimeType ?? 'application/octet-stream'` fallback means the absent-mime + oversized-blob path is entirely unverified. Consider adding a sibling test that sends an oversized blob via `embeddedContext` (a resource with `blob` and no `text`) and asserts the model receives a text placeholder rather than `inlineData`. Cover both an explicit `mimeType` (e.g., `audio/wav`) and missing `mimeType` (falls back to `application/octet-stream`).

_— qwen3.7-max via Qwen Code /review_

Comment (doudouOUC) [reply packages/cli/src/acp-integration/session/Session.ts:2961]: Thanks — deferring this one. The core clamp logic is exhaustively unit-tested in `inlineMediaLimit.test.ts` (thresholds, mime types, env-var override, sanitization). The existing integration test proves the Session wiring pattern works end-to-end. The blob branch calls the same `clampInlineMediaPart` — the only unique line is the `mimeType ?? 'application/octet-stream'` fallback, which doesn't warrant a full integration test in this PR's scope. Happy to add it as a follow-up if the blob path gets more complex.

**Comment by wenshao** [comment]
## Verification Report

**Branch**: `feat/daemon-inline-media-clamp`
**Base**: `daemon_mode_b_main`
**Environment**: macOS Darwin 25.4.0, Node.js v22.17.0

### Test Results

| Check | Command | Result |
|-------|---------|--------|
| Unit Tests | `vitest run packages/core/src/core/inlineMediaLimit.test.ts` | ✅ 12 tests passed (3ms) |
| Session Integration | `vitest run packages/cli/src/acp-integration/session/Session.test.ts` | ✅ 85 tests passed (392ms) |
| Core Type Check | `npm run typecheck --workspace=packages/core` | ✅ Clean |
| CLI Type Check | `npm run typecheck --workspace=packages/cli` | ✅ Clean (0 errors) |
| Core Build | `npm run build --workspace=packages/core` | ✅ Success |
| ESLint | `eslint` on all changed files | ✅ No errors |
| Whitespace | `git diff --check` | ✅ Clean |

### Code Review Notes

- `approxBase64Bytes()` correctly estimates decoded size from base64 length without allocating a decode buffer — efficient for the hot prompt path
- `clampInlineMediaPart()` properly handles all Part variants (inlineData, text, fileData) and passes non-media parts through untouched
- Mime type sanitization via `sanitizeMimeForPlaceholder()` prevents injection through untrusted resource/MCP server-provided mime types (test confirms no newlines or bracket sequences leak)
- `QWEN_CODE_MAX_INLINE_MEDIA_BYTES` env var override has proper validation (positive, finite, floor'd)
- The placeholder message is user-actionable: suggests resize/compress or @file path as alternatives
- `audio: true` in HTTP dispatch is consistent with acpAgent.ts since `#resolvePrompt` handles audio identically to image

### Note

Initial test run hit a workspace symlink staleness issue (not PR-related). After `npm install --ignore-scripts` to refresh workspace links, all tests passed cleanly.

### Verdict

**✅ Ready to merge** — all test plan items pass, implementation is well-tested with 12 dedicated unit tests and 1 integration test covering the clamping path.

---
*Verified by wenshao*

File: packages/cli/src/acp-integration/session/Session.ts [hunk 3]
```diff
@@ -2853,12 +2854,12 @@ export class Session implements SessionContext {
           return { text: part.text };
         case 'image':
         case 'audio':
-          return {
+          return clampInlineMediaPart({
```
Comment (wenshao) [inline packages/cli/src/acp-integration/session/Session.ts:2857]: **[Suggestion]** `clampInlineMediaPart` is wired at two call sites (image/audio branch here, blob branch at ~line 2960), but a third `inlineData`-producing path is unguarded: the `readManyFiles` result at ~line 2935. Non-string `contentParts` from `readManyFiles` (binary files like images/PDFs read via `@file`) are pushed directly into `processedQueryParts` without clamping. A user `@file`-referencing a 20 MB PNG would bypass the very protection this PR introduces.

```suggestion
          return clampInlineMediaPart({
            inlineData: {
              mimeType: part.mimeType,
              data: part.data,
            },
          });
```

Consider also wrapping the `readManyFiles` path:
```ts
for (const part of contentParts) {
  if (typeof part === 'string') {
    processedQueryParts.push({ text: part });
  } else {
    processedQueryParts.push(clampInlineMediaPart(part));
  }
}
```

_— qwen3.7-max via Qwen Code /review_

Comment (doudouOUC) [reply packages/cli/src/acp-integration/session/Session.ts:2857]: Agreed, good catch — fixed in 3077ab6. The readManyFiles path was indeed unguarded.

**Comment by doudouOUC** [comment]
## Review round 2 summary

| Thread | Author | Action | Commit |
|--------|--------|--------|--------|
| readManyFiles @file path unguarded (Session.ts:2857) | wenshao | **Fixed** — wrapped non-string contentParts with `clampInlineMediaPart` | 3077ab6 |

Resolved 1/1 threads.

**Review by wenshao — APPROVED** [review]

File: packages/core/src/core/inlineMediaLimit.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  mimeType: string,
+  bytes: number,
+  limitBytes: number,
+): string {
+  // Sanitize: the mime can originate from an untrusted resource/MCP server,
+  // and is embedded into a bracketed envelope the model reads as text.
+  const mime = sanitizeMimeForPlaceholder(mimeType);
+  return (
+    `[Media omitted: ${mime} is ~${formatMb(bytes)}MB, exceeding the ` +
+    `${formatMb(limitBytes)}MB inline limit. Ask the user to resize/compress ` +
```
Comment (wenshao) [inline packages/core/src/core/inlineMediaLimit.ts:74]: **[Suggestion]** The placeholder text advises "reference it via an @file path so it can be read from disk" but this is misleading for 2 of the 3 call sites:

- `Session.ts:2939` (readManyFiles): the user already used `@file` — the advice is redundant.
- `Session.ts:2961` (MCP blob): the data comes from an MCP resource with no local file — the advice is nonsensical.

Only the direct user-input path (image/audio) benefits from this suggestion. Consider removing the @file advice from the generic placeholder — "Ask the user to resize/compress it" is universally correct.

```suggestion
  return (
    `[Media omitted: ${mime} is ~${formatMb(bytes)}MB, exceeding the ` +
    `${formatMb(limitBytes)}MB inline limit. Ask the user to resize or ` +
    `compress the file before attaching.]`
  );
```

_— claude-opus-4-6 via Qwen Code /review_

File: packages/core/src/core/inlineMediaLimit.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+function formatMb(bytes: number): string {
+  return (bytes / (1024 * 1024)).toFixed(1);
+}
+
+/**
+ * Build the placeholder text substituted for an oversized inline media part.
+ */
+export function oversizedMediaPlaceholder(
```
Comment (wenshao) [inline packages/core/src/core/inlineMediaLimit.ts:64]: **[Suggestion]** `oversizedMediaPlaceholder` is exported and re-exported through the barrel (`export * from './core/inlineMediaLimit.js'`) but has zero external consumers — it is only called by `clampInlineMediaPart` in this same file. This exposes an implementation detail (including the exact string format) as public API surface, constraining future refactoring of placeholder wording.

```suggestion
function oversizedMediaPlaceholder(
```

_— claude-opus-4-6 via Qwen Code /review_

File: packages/core/src/core/inlineMediaLimit.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+): Part {
+  const data = part.inlineData?.data;
+  if (!data) {
+    return part;
+  }
+  const bytes = approxBase64Bytes(data);
+  if (bytes <= limitBytes) {
+    return part;
+  }
+  const mimeType = part.inlineData?.mimeType ?? 'application/octet-stream';
```
Comment (wenshao) [inline packages/core/src/core/inlineMediaLimit.ts:96]: **[Suggestion]** The `mimeType ?? 'application/octet-stream'` fallback is untested — every test case provides an explicit `mimeType`. Since the `Part` type makes `mimeType` optional, real callers (e.g., the blob path with a missing `contextPart.mimeType`) can hit this branch.

```typescript
it('falls back to application/octet-stream when mimeType is missing', () => {
  const part = { inlineData: { data: 'A'.repeat(2000) } };
  const result = clampInlineMediaPart(part as Part, 1000);
  expect(result.text).toContain('application/octet-stream');
  expect(result.text?.toLowerCase()).toContain('omitted');
});
```

_— claude-opus-4-6 via Qwen Code /review_

**Review by chiga0 — APPROVED** [review]
## QoderWork Code Review — PR #4646

**Decision: APPROVED** ✅

### Summary
`feat(daemon): clamp oversized inline media on the prompt path` — prevents oversized inline image/audio/blob payloads from inflating daemon request size and token budget.

### Review Notes
- **`clampInlineMediaPart`**: Clean pure-function guard on individual `Part` objects. Returns text placeholder for oversized media, passes through unchanged otherwise.
- **`approxBase64Bytes`**: O(1) estimation via string length arithmetic — no decode/copy, safe on multi-MB payloads on the prompt hot path. Correctly handles empty input, padding (`==`, `=`), and `data:` URL prefix.
- **Wiring completeness**: All 3 inline media entry points in `#resolvePrompt` covered — direct image/audio parts, `@file` readManyFiles binary parts, and blob embedded context.
- **Security**: Mime type in the placeholder text is sanitized via `sanitizeMimeForPlaceholder` against prompt injection (tested with `[SYSTEM: hijack` payload).
- **`dispatch.ts`**: `audio: true` correctly mirrors actual `#resolvePrompt` handling (both image and audio become `inlineData` Parts).
- **`fileUri!` assertion**: Correct — only applies on `resource_link` path where `clampInlineMediaPart` returns the original part unchanged.
- **Env-configurable ceiling**: `QWEN_CODE_MAX_INLINE_MEDIA_BYTES` with 10MB default. Lazy-evaluated at each call, so runtime env changes take effect immediately.
- **Test coverage**: 8 unit tests + 1 integration test covering base64 sizing, env override edge cases, clamping behavior, mime sanitization, and end-to-end Session prompt degradation.

No findings. LGTM.

---
_This review was generated by QoderWork AI_
